### PR TITLE
[WIP] RSP-1367: Tooltip Warmup/Cooldown logic

### PR DESCRIPTION
**Pull request by pst67662**
Closes 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [Issue](https://github.com/adobe-private/react-spectrum-v3/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-stately/tooltip/src/TooltipManager.ts [hunk 1]
```diff
@@ -10,67 +10,76 @@
...
 
-export class TooltipManager {
+const CLOSE_DELAY = 220; // --spectrum-global-animation-duration-400
+const WARMUP_DELAY = 350;  // --spectrum-global-animation-duration-700
+const COOLDOWN_DELAY = 400; // --spectrum-global-animation-duration-1000
 
-  visibleTooltip?: null | {triggerId: string, state: {open: boolean, setOpen(value: boolean), tooltipManager: string}}
-  hoverHideTimeout?: null | ReturnType<typeof setTimeout>;
-  hoverShowTimeout?: null | ReturnType<typeof setTimeout>;
+let visibleTooltip: null | TooltipState = null;
```
Comment (snowystinger) [inline packages/@react-stately/tooltip/src/TooltipManager.ts:19]: you should be able to keep these inside the class TooltipManager

then, to make the entire class a protected singleton you can do this

```
class TooltipManager implements MultipleTolltipManager {...}
let tooltipManager = new TooltipManager();
export {tooltipManager as TooltipManager}
```


Another approach you could take is to leave these out as they are, get rid of the class entirely and follow a reducer pattern
https://github.com/adobe-private/react-spectrum-v3/blob/398de48a8b5d653f699f0a993df42734fbe2c9a8/packages/%40react-stately/data/src/useAsyncList.ts#L79

**Comment by dannify** [comment]
I haven't gone into the code too much but testing out the storybook, it looks good and the logic seems to be what we want. Nice work

File: packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js [hunk 2]
```diff
@@ -23,7 +23,7 @@ let theme = {
   medium: scaleMedium
 };
 
-describe('TooltipTrigger', function () {
+describe.skip('TooltipTrigger', function () {
```
Comment (LFDanLu) [inline packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js:26]: Will these be fixed in this pull? If not, mind adding a TODO comment just so it is easier to look up in the future?

File: packages/@react-stately/tooltip/src/TooltipManager.ts [hunk 3]
```diff
@@ -10,67 +10,76 @@
...
+  openTooltip() {
     // Close previously open tooltip
-    if (this.visibleTooltip) {
-      this.visibleTooltip.state.setOpen(false);
+    if (visibleTooltip) {
+      visibleTooltip.close();
     }
+
+    this.state.open();
+    visibleTooltip = this.state;
```
Comment (LFDanLu) [inline packages/@react-stately/tooltip/src/TooltipManager.ts:43]: Noticed that it is possible to have multiple tooltips open at once through the following flow:
1. Hover over the copy button to trigger a tooltip opening
2. Press "Tab" to focus another button. Note that the previous tooltip closes and the newly focused button shows its tooltip (this is correct behavior so far).
3. Hover over a different button. Note that two tooltips are now showing

I'll need to dig into why this is happening, but noting it here for now

File: packages/@react-stately/tooltip/test/TooltipManager.test.js [hunk 4]
```diff
@@ -12,7 +12,7 @@
 
 import {TooltipManager} from '../';
 
-describe('TooltipManager', () => {
+describe.skip('TooltipManager', () => {
```
Comment (LFDanLu) [inline packages/@react-stately/tooltip/test/TooltipManager.test.js:15]: Same as above.

File: packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js (hunk 2 above)
Comment (pst67662) [reply packages/@react-spectrum/tooltip/test/TooltipTrigger.test.js:26]: tests will be updated in this PR, skipped for initial commit for build to pass

File: packages/@react-stately/tooltip/src/TooltipManager.ts (hunk 1 above)
Comment (pst67662) [reply packages/@react-stately/tooltip/src/TooltipManager.ts:19]: need little help with this

1. if TooltipManager is singleton then there is a problem with passing state correctly. I'll need to pass state in useTooltipTrigger eveytime when TooltipManager is called

something like:
```
 function useTooltipTrigger(props, state) {
  ...
  let {pressProps} = usePress({
    onPress: () => state.tooltipManager.closeTooltip(state)
  });
}
```
would this be OK?

Comment (pst67662) [reply packages/@react-stately/tooltip/src/TooltipManager.ts:19]: 2. I got lost with `reducer`, is there a proper way to handle timeout inside of `reducer`?
```
    warmupTimer = setTimeout(() => {
      this.openTooltip(); <- returns state {isOpen: true}
      warmedup = true;
    }, WARMUP_DELAY);
```

File: packages/@react-stately/tooltip/src/TooltipManager.ts (hunk 3 above)
Comment (pst67662) [reply packages/@react-stately/tooltip/src/TooltipManager.ts:51]: good catch...  I looked into, but didnt find real reason why it happens.... maybe to open separate ticket for this one

**Comment by snowystinger** [comment]
Was incorporated into https://github.com/adobe/react-spectrum/pull/843

**Comment by snowystinger** [comment]
Thank you @pst67662 !!!!

**Pull request closed without merging**